### ISSUE TITLE
Pack maps no Nemotron-H tensor class beyond the routed-expert stacks

### ISSUE THREAD

**Issue by Alberto-Codes** [Alberto-Codes/vramfit#183]
## What

`vramfit pack` maps two group shapes after #180: a decoder-layer
group under any naming family, and a routed-expert stack. Every
other `stack` group still raises a `PackError` that names it.

On the Nemotron 3.5 Lightning target a `--group-by stack` scan
groups every 2-dimensional float parameter, so the recipe also
carries:

- `backbone.layers.<n>.mixer.in_proj` and `.out_proj` — Mamba-2.
- `backbone.layers.<n>.mixer.{q,k,v,o}_proj` — the 6 attention
  layers.
- The router and the shared experts.

So a `layer`-keyed recipe packs the target today. A whole-model
`stack`-keyed recipe does not.

## Why this is a decision, not a task

ADR-0022 decision 1 fixes the HF-to-GGUF class table at the seven
projections of a llama-family layer. Extending it to Nemotron-H
needs two rulings:

1. The class table for this family. #159 read the names from a
   shipped artifact: `ssm_in`, `ssm_out`, `attn_q`, and the rest.
2. What a recipe does with a group llama-quantize never
   quantizes. #159 lists `ssm_conv1d.weight`, `ffn_gate_inp.weight`,
   `ssm_a`, `ssm_d`, `ssm_norm.weight`. A scan measures them and
   the solver prices them, and the pack cannot spend those bits.
   An override for one of them is a silent no-op today.

Point 2 is the sharp one. It moves predicted bytes away from
packed bytes, which ADR-0012 decision 4 re-checks.

## Blocks

Corrected 2026-08-15. This section read "Nothing yet" and that is no
longer true. Two issues now wait on this ruling, both as wired edges:

- **#204** — the scan prices 23 `mixer.conv1d.weight` cells the
  quantizer never touches. Point 2 below rules what a recipe does with
  a group `llama-quantize` never quantizes, so #204 cannot be fixed
  without pre-empting it.
- **The dense side of the chart #158 budget.** Chart Notes record that
  a 30B pack completes only with every Mamba layer at nominal 8 until
  this ticket rules the `ssm_in` class (#228, ADR-0028). Measured from
  the #229 gate packs, that pin costs **1,101,373,440 B** — the
  difference between 3 and 10 expert stacks escaping Q2_0 inside a
  10.5 GiB weight budget. Arithmetic on
  [#249](https://github.com/Alberto-Codes/vramfit/issues/249#issuecomment-5300752158).

A `layer`-keyed recipe still reaches the chart #158 Destination. This
ticket lifts the pack to the granularity #161 made measurable.

**Comment by Alberto-Codes** [comment]
This gap is wider than Nemotron-H.

Alibaba released Qwen3.8-27B on 2026-08-14. It is dense, so it carries no
routed-expert stacks at all. Its 64 layers repeat three `linear_attention`
layers to one `full_attention` layer, and the linear layers are Gated DeltaNet
with `linear_conv_kernel_dim: 4`. It also ships a native vision encoder at
depth 27 and hidden size 1152.

Two consequences for the tensor-class map:

- The linear-attention conv1d and state tensors appear in a second
  architecture. See #204, which reports the scan side of the same gap.
- The vision tower adds patch embeddings, its own attention blocks, and a
  projector. #236 asks whether a non-text encoder enters the map or the weight
  budget at all. That is a decision and it blocks any map entry for those
  tensors.

#235 parks Qwen3.8-27B as a target and records the architecture facts. Nothing
here changes this issue's scope for the current north star. It argues for a
map keyed on tensor role rather than on checkpoint family.

**Comment by Alberto-Codes** [comment]
## Sequencing question, raised from #249

This ticket now sits in front of a measurable amount of budget. Noting
it here so the sequencing is tracked rather than assumed.

Chart #158's Notes record that a 30B pack completes only with every
Mamba layer at nominal 8 until this ticket rules the `ssm_in` class
(#228, ADR-0028). The #229 gate ran that way. All three gate packs
carry an identical non-expert block of **2,374,762,400 B**.

What that costs, derived from the gate packs:

- Dropping every quantizable non-expert group from nominal 8 to
  nominal 4 frees **1,101,373,440 B**.
- Against a 10.5 GiB weight budget, that is **0.309 more bits per
  weight** available across the 29.375B of routed experts.
- It raises the count of expert stacks that escape Q2_0 from **3 of
  46 to 10 of 46**.

The full arithmetic is in the
[#249 comment](https://github.com/Alberto-Codes/vramfit/issues/249).

So this ticket is not only a granularity question for a `stack`-keyed
recipe. It gates about 1.03 GiB of the weight budget on the target the
chart is aimed at.

**The question, for the maintainer:** does this ticket rule before the
campaign spends again, or does the next spend run with the non-expert
block pinned at nominal 8?

This comment rules nothing and adds no scope to the ticket.

**Comment by Alberto-Codes** [comment]
## Correction: the sequencing numbers are keyed to a superseded budget

The [sequencing comment](https://github.com/Alberto-Codes/vramfit/issues/183#issuecomment-5300796529) above prices this ticket against the 10.5 GiB weight budget. **#257 superseded that budget on 2026-08-15.** The card is 16 GiB and the weight budget is 14.5 GiB.

The freed bytes do not change. **1,101,373,440 B** still separates a non-expert block at nominal 8 from one at nominal 4. What changes is what those bytes buy.

| budget | dense nominal 8 | dense nominal 4 |
|--------|-----------------|-----------------|
| 10.5 GiB (superseded) | 3 of 46 | 10 of 46 |
| **14.5 GiB (current)** | **27 of 46** | **33 of 46** |

So this ticket is worth **6 more expert stacks escaping `Q2_0`**, not 7 — and it now sits against a base of 27 rather than 3. The packed sizes, from the same byte identity:

- dense nominal 8, 27 stacks at `Q4_0`: 15,485,665,184 B = 14.422 GiB, 79.7 MiB under budget.
- dense nominal 4, 33 stacks at `Q4_0`: 15,461,900,192 B = 14.400 GiB, 102.4 MiB under budget.

**One smaller correction.** The comment states the relief is worth "0.309 more bits per weight". I get 0.300:

```
1,101,373,440 B x 8 = 8,810,987,520 bits
8,810,987,520 / 29,374,808,064 params = 0.29995 bits per weight
```

That quantity is budget-independent, so it should not have moved with the card. I may be missing the intended denominator — flagging rather than asserting.

## What this does to the ticket's priority

It cuts both ways, and the ticket's own framing no longer decides it.

At 10.5 GiB the relief was the difference between a recipe that relieved almost nothing (3 of 46) and one that relieved a little (10 of 46). That made this ticket close to load-bearing.

At 14.5 GiB a recipe already relieves 27 of 46 without this ticket, and chart #158 states that **a `layer`-keyed recipe still reaches the Destination**. So the ruling is now an optimization worth 6 stacks rather than a gate.

It still sets the operating point any measured arm packs against. A measured arm packed at nominal 8 sits 6 stacks off the eventual point, and a later ruling here makes it a re-buy. That trade belongs on #249, which rules the spend, and it is raised there.

Correcting the record only. This comment rules nothing and does not claim the ticket.

Same rot class as #277, which tracks ADR-0021 decision 4 arguing from the same superseded budget.

**Comment by Alberto-Codes** [comment]
## Ruling here now costs a re-buy, from #249 (2026-08-16)

#249 ruled the campaign packs at **dense nominal 8**, which relieves 35
of 46 expert stacks to `Q4_0`. It did not wait on this ticket. Record:
[chart #158 Notes](https://github.com/Alberto-Codes/vramfit/issues/158).

If this ticket later relieves the quantizable dense classes to nominal 4,
the figure becomes **41 of 46** and every packed arm in
[#300](https://github.com/Alberto-Codes/vramfit/issues/300) is a re-buy.

That re-buy is four packs plus four evaluations on a cold pod, so it
re-pays the 15-minute f16 convert and the base-logits pass. About 29
minutes and $1.61.

So this ticket is worth 6 expert stacks, and after #300 runs it is also
worth one more pod. Ruling it before #300 books saves that.

**Comment by Alberto-Codes** [comment]
## Narrow ruling recorded, 2026-08-16, and this ticket stays open

**Maintainer ruling: the #300 campaign packs dense groups at nominal 8.**
Given before the pod booked, so the campaign runs once rather than twice.

That ruling settles the campaign's operating point only. It does not reach
this ticket's question, which is the Nemotron-H class table and what a
recipe does with a group `llama-quantize` never quantizes. **This ticket
stays open on that wider question.**

Record: [chart #158 Notes](https://github.com/Alberto-Codes/vramfit/issues/158).

Recorded here so a later session does not read the campaign's operating
point as an answer to the class-table question.

**Comment by Alberto-Codes** [comment]
## Measured against a whole-model recipe, 2026-08-18

PR #360 builds ADR-0029, so `vramfit plan --checkpoint` now emits a
210-group `stack`-keyed recipe for this target. That is the input this
ticket predicted. `types.tensor_overrides` refuses it, and the first
three refusals read:

```
group "model.embeddings" has no GGUF tensor mapping
group "model.layers.0.mixer.in_proj" has no GGUF tensor mapping
group "model.layers.1.mixer.shared_experts.up_proj" has no GGUF tensor mapping
```

Two facts the class-table ruling should carry.

**The embedding name is a fourth case, not part of the class table.**
`EMBEDDING_GROUPS` holds `model.embed_tokens` and
`backbone.embeddings`. The scan emits **`model.embeddings`** on this
target, which is in neither. Both maps this project holds carry that
name, so the embedding group has never mapped here. It reaches
`tensor_overrides` and refuses as an unmapped group rather than
driving `--token-embedding-type`. That is a one-entry fix and it is
independent of the Mamba-2 and attention classes.

**The count is 164 groups, not a handful.** A 46-group map leaves 164
uncovered on the 210-group discovery, and every one of them is a dense
group this ticket covers. So the class table is what stands between a
planned whole-model recipe and a packed one.

#360 does not touch pack. It documents the gap in
`docs/reference/cli.md` and points here.

**Comment by Alberto-Codes** [comment]
## Prep for the ruling: both halves priced from primary sources

Measured this session. The class table comes from llama.cpp at
`3653e6d6d` (b10326), which ADR-0027 makes part of the frame. The byte
figures come from the checkpoint's own safetensors shard headers at
`~/models/NVIDIA-Nemotron-3.5-Lightning-30B-A3B-BF16`.

### Point 1: the class table is confirmed, and the pack table needs a root too

`gguf-py/gguf/tensor_mapping.py` at `3653e6d6d` carries every row of
the nine-name table in the
[2026-08-12 comment](https://github.com/Alberto-Codes/vramfit/issues/183#issuecomment-5285297739),
each tagged `nemotron-h` or `nemotron-h-moe`, each rooted
`backbone.layers.{bid}.mixer.`. The table survives at the instrument
commit.

Three facts the pack side must absorb beyond the names:

1. `_LAYER_TENSOR` reads `^model\.layers\.` and the checkpoint roots at
   `backbone.layers.`.
2. `_LAYER_TENSOR` matches two dot-separated segments. The shared-expert
   suffix `mixer.shared_experts.down_proj` carries three.
3. #360 emits `model.` for every group (#362, gap 1). So the table must
   state which root it keys on, and the two tickets agree or the recipe
   still refuses.

What the classes weigh, at bf16 from the shard headers:

| class | GGUF stem | n | params | MiB |
|---|---|---:|---:|---:|
| `mixer.in_proj` | `ssm_in` | 23 | 637,034,496 | 1215.05 |
| `mixer.out_proj` | `ssm_out` | 23 | 253,231,104 | 483.00 |
| `mixer.shared_experts.up_proj` | `ffn_up_shexp` | 23 | 229,490,688 | 437.72 |
| `mixer.shared_experts.down_proj` | `ffn_down_shexp` | 23 | 229,490,688 | 437.72 |
| `mixer.{q,k,v,o}_proj` | `attn_*` | 6 each | 140,378,112 | 267.75 |
| **total** | | | **1,489,625,088** | **2841.24** |

That is 2.775 GiB the pack pins at nominal 8 today.

### Point 2 is worth 16.83 MiB, not a budget line

`tensor_allows_quantization` at `src/llama-quant.cpp:289-364` refuses
this target's five classes through three filters, none of them a type
table:

- `ggml_n_dims(tensor) < 2` drops `mixer.A_log`, `mixer.D`,
  `mixer.dt_bias`, `mixer.conv1d.bias`, and
  `mixer.gate.e_score_correction_bias`.
- `name.find("_norm.weight")` drops `mixer.norm.weight`,
  `norm.weight`, and `backbone.norm_f.weight`.
- `name.find("ffn_gate_inp.weight")` drops `mixer.gate.weight`.
- `name.find("ssm_conv1d")` drops `mixer.conv1d.weight`.

Only two of the five carry 2 or more dimensions, so only two reach a
size source that filters on rank:

| class | ndim | shape | n | params | MiB |
|---|---:|---|---:|---:|---:|
| `mixer.gate.weight` | 2 | [128, 2688] | 23 | 7,913,472 | 15.09 |
| `mixer.conv1d.weight` | 3 | [6144, 1, 4] | 23 | 565,248 | 1.08 |
| **total** | | | | **8,822,784** | **16.83** |

**16.83 MiB is 0.104 % of the 15.776 GiB budget.** So point 2 moves
predicted bytes away from packed bytes by a rounding error. It is an
honesty question about a silent no-op, and never a budget question.
The sequencing comment's 1,101,373,440 B belongs entirely to point 1.

One further fact for point 2. llama.cpp exports the predicate as
`llama_quant_tensor_allows_quantization` at `src/llama-ext.h:48`. It
takes `quantize_state_impl *`, an internal struct, and `llama-ext.h`
sits under `src/` and not `include/`. So no CLI reaches it, and a
vramfit-side rule copies the list rather than calling it. #207 asks how
this project marks a test that pins a copied external contract.

This comment rules nothing.

**Comment by Alberto-Codes** [comment]
## Correction to the prep comment: the ruling carries three items, not two

The [prep comment](https://github.com/Alberto-Codes/vramfit/issues/183#issuecomment-5349094596)
above omits the embedding case. The
[2026-08-19 comment](https://github.com/Alberto-Codes/vramfit/issues/183#issuecomment-5348159321)
already records it, measured against the 210-group recipe #360 emits.

`EMBEDDING_GROUPS` holds `model.embed_tokens` and
`backbone.embeddings`. The scan emits `model.embeddings`, which is in
neither, so the embedding group has never mapped on this target. It
refuses as an unmapped group rather than driving
`--token-embedding-type`.

It weighs 352,321,536 parameters, at 672.00 MiB at bf16 from the shard
headers. That is 4.16 % of the 15.776 GiB budget, and the largest
single group in the model outside the expert stacks. The prep comment's
2.775 GiB total for point 1 excludes it, because the
[2026-08-12 comment](https://github.com/Alberto-Codes/vramfit/issues/183#issuecomment-5285297739)
scopes that table to layer classes.

So the ruling carries three items:

1. The layer class table, worth 2.775 GiB.
2. The embedding entry, worth 672.00 MiB, and a one-entry fix.
3. What a recipe does with an unquantizable group, worth 16.83 MiB.

The prep comment's byte figures stand. Only its item count was wrong.

**Comment by Alberto-Codes** [comment]
## Ruled 2026-08-20, and recorded

Maintainer ruling, 2026-08-20, live session. Four items:

1. **Class table.** The pack class table gains the nine Nemotron-H
   rows under a free prefix, the 2026-08-12 amendment's shape. The
   two-root refusal stands.
2. **Type table.** A mapped class whose rows refuse the 256
   super-block drives through ADR-0028's table. This answers the
   hand-off in ADR-0028's Consequences, and `ssm_in`'s wait ends.
   Peer review on PR #369 surfaced the gap.
3. **Embedding.** The embedding name set gains `model.embeddings`,
   which is worth 672.00 MiB.
4. **F16 pin.** An unquantizable class pins at the F16 passthrough.
   `plan` assigns it, the solver never prices lower, and pack
   refuses a lower width. The filter list is a copied external
   contract, and #207 carries the test-marker question.

Record: the 2026-08-20 amendments to
[ADR-0012](https://github.com/Alberto-Codes/vramfit/blob/main/docs/adr/0012-gguf-type-mapping.md)
and
[ADR-0028](https://github.com/Alberto-Codes/vramfit/blob/main/docs/adr/0028-expert-stack-type-table.md)
(PR #369). #368 lands the build.
